test/plugin_functional/test_suites/core_plugins/http_versioned.ts

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import expect from '@kbn/expect';
+import '@kbn/core-provider-plugin/types';
+import { PluginFunctionalProviderContext } from '../../services';
+
+export default function ({ getService }: PluginFunctionalProviderContext) {
+  const supertest = getService('supertest');
+
+  // routes defined in the `core_http` test plugin
+  describe('Route version negotiation', () => {
+    describe('internal route', () => {
+      it('serves requests with the handler corresponding to the specified version', async () => {
+        await supertest
+          .get('/api/core_http/internal_versioned_route')
+          .set('Elastic-Api-Version', '1')
+          .expect(200)
+          .then((response) => {
+            expect(response.body).to.eql({
+              version: 1,
+            });
+          });
+        await supertest
+          .get('/api/core_http/internal_versioned_route')
+          .set('Elastic-Api-Version', '2')
+          .expect(200)
+          .then((response) => {
+            expect(response.body).to.eql({
+              version: 2,
+            });
+          });
+      });
+
+      it('returns the version in response headers', async () => {
+        await supertest
+          .get('/api/core_http/internal_versioned_route')
+          .set('Elastic-Api-Version', '1')
+          .expect(200)
+          .then((response) => {
+            expect(response.headers['elastic-api-version']).to.eql('1');
+          });
+        await supertest
+          .get('/api/core_http/internal_versioned_route')
+          .set('Elastic-Api-Version', '2')
+          .expect(200)
+          .then((response) => {
+            expect(response.headers['elastic-api-version']).to.eql('2');
+          });
+      });
+
+      it('requires the version header to be set', async () => {
+        await supertest
+          .get('/api/core_http/internal_versioned_route')
+          .unset('Elastic-Api-Version')
+          .expect(400)
+          .then((response) => {
+            expect(response.body.message).to.match(/Please specify.+version/);
+          });
+      });
+    });
+
+    describe('public route', () => {
```

Comment:
Maybe also a test confirming we don't need a version to be specified for public routes?

Reply:
Yeah, I added it initially, but that behavior changes between prod and dev mode (the test passes in production mode given your statement is true there, but not on local mode given we're doing strict validation in that case).

The evil of non-configurable, per-env behavior divergence 😞